### DISCUSSION TITLE
Set authState after user signs in?

### DISCUSSION THREAD

**Discussion by chalupabatman07** [urql-graphql/urql#2246] [Q&A]
Hey all, I am pretty new with urql and have been liking it, but I seem to be a bit confused about what I am suppose to do with the authExchange or I am probably doing something wrong? I am not sure. I am creating a React Native app where the user lands on a lander they navigate to sign up where they enter their phone number and then enter their verification code which then we create the user and get the JWT Token. Here's the client setup I have:

```
const client = createClient({
    url: 'http://10.0.2.2:8080/graphql',
    exchanges: [
      dedupExchange,
      cacheExchange,
      authExchange<AuthState>({
        getAuth: async ({ authState }) => {
          console.log('begin getAuth');
          if (!authState) {
            const token = await SecureStore.getItemAsync('token');

            if (token) {
              console.log('token found');
              return { token };
            }
            console.log('token not found');
            return null;
          }
          return null;
        },
        addAuthToOperation: ({ authState, operation }) => {
         console.log('begin addAuthToOperation');
          if (!authState || !authState.token) {
            console.log('no authState or no authState.token');
            return operation;
          }

          console.log('token or authState found');

          const fetchOptions =
            typeof operation.context.fetchOptions === 'function'
              ? operation.context.fetchOptions()
              : operation.context.fetchOptions || {};

          return makeOperation(operation.kind, operation, {
            ...operation.context,
            fetchOptions: {
              ...fetchOptions,
              headers: {
                ...fetchOptions.headers,
                Authorization: `Bearer ${authState.token}`,
              },
              credentials: 'include',
            },
          });
        },
      }),
      fetchExchange,
    ],
  });
  ```
  Pretty much followed the documents here: https://formidable.com/open-source/urql/docs/api/auth-exchange/
  
  On the phone verification screen they pretty much enter your verification code and click continue which runs this method:
```  
const CreateUser = `
  mutation createUser($phoneNumber: String!) {
    createUser(phoneNumber: $phoneNumber) {
      token
    }
  }
`;

const [, createToken] = useMutation(CreateUser);

const handleOnContinue = async () => {
    const result = await createUser({ phoneNumber }); // Mutation to create user returns a token
    const token = result.data.createUser.token;
    await SecureStore.setItemAsync('token', token); // Adding token to securestore
    navigation.navigate(MainRoutes.EmailRegistration); // Navigates to email registration
  };
```

Here's the output for my console:
```
begin getAuth
token not found
begin addAuthToOperation
no authState or no authState.token
begin addAuthToOperation
no authState or no authState.token
begin addAuthToOperation
no authState or no authState.token
begin addAuthToOperation
no authState or no authState.token
begin addAuthToOperation
no authState or no authState.token
begin addAuthToOperation
no authState or no authState.token
```
I am not sure if I am missing something, but how do I save a token after the user signs up and get the `authState` to update with the token after the mutation call? Seems like my `getAuth` runs initially when the app first opens and I create my client, but I feel like I did something wrong. Unless my `getAuth` is only initially ran once

**Comment by drj17** [comment]
I had a very similar question and the way I've seen some open source projects handle is it by re-initializing the urql client when the logged in state changes:

```
  const client = useMemo(() => {
    return createClient({
      url: GRAPHQL_API,
      exchanges: [
        devtoolsExchange,
        dedupExchange,
        cacheExchange({ resolvers, updates, keys }),
        authExchange({
          getAuth,
          addAuthToOperation,
          willAuthError,
          didAuthError,
        }),
        fetchExchange,
      ],
    })
  }, [state.isLoggedIn])
```

**Comment by JoviDeCroock** [comment]
Hey and welcoem to urql!

There are a few ways around this, you can create the urql-client and authExchange in the front-end components and react to auth-changes that way, this can be inconvenient as it will reset your cache. Generally I advise people to use an `event-emitter` between their front-end and the global js-scope in which the client is defined, as answered [here](https://github.com/urql-graphql/urql/discussions/2647#discussioncomment-3501919).
